File: tests/bindings/e2e/cpp/worker_process_fixture.cpp
#include "common/memory/backend.hpp"
#include "common/runtime/workflow_store.hpp"
#include "worker/worker_process.hpp"
#include "worker/worker_runtime.hpp"
#include "worker/workflow_step_handlers.hpp"

#include <atomic>
#include <chrono>
#include <stdexcept>
#include <thread>

class ProcessWorkflowStepHandler final : public statespec_generated::worker::IWorkflowStepHandler
{
  public:
    void handle_validate_request(
        const statespec_generated::worker::WorkflowStepHandlerContext& context
    ) override
    {
        if (context.workflow_name != "ProvisionService" || context.step_name != "validate_request")
        {
            throw std::runtime_error("unexpected workflow step");
        }
        handled_validate_request = true;
    }

    void handle_create_remote_service(
        const statespec_generated::worker::WorkflowStepHandlerContext&
    ) override
    {
        handled_create_remote_service = true;
    }

    void handle_wait_for_remote_service(
        const statespec_generated::worker::WorkflowStepHandlerContext&
    ) override
    {
        handled_wait_for_remote_service = true;
    }

    std::atomic_bool handled_validate_request = false;
    std::atomic_bool handled_create_remote_service = false;
    std::atomic_bool handled_wait_for_remote_service = false;
};

int main()
{
    statespec::backend::memory::InMemoryBackend backend;
    statespec_generated::worker::WorkerRuntime runtime{backend};
    ProcessWorkflowStepHandler handler;
    statespec_generated::worker::WorkerProcessConfig config;
    config.worker_poll_interval_ms = 1;
    statespec_generated::worker::WorkerProcess process{runtime, handler, config};

    if (process.join() == 0)
    {
        throw std::runtime_error("joining a WorkerProcess before start should fail");
    }
    process.request_stop();

    runtime.bootstrap();
    statespec::backend::runtime::RuntimeWorkflowStore workflows;
    workflows.start(
        backend, statespec::backend::StartWorkflowRequest{
                     "wf-process-1",
                     "ProvisionService",
                     1,
                     "validate_request",
                     statespec::backend::Json::object({}),
                 }
    );

    if (process.start() != 0)
    {
        throw std::runtime_error("starting WorkerProcess failed");
    }
    if (process.start() == 0)
    {
        throw std::runtime_error("starting WorkerProcess twice should fail");
    }
    std::this_thread::sleep_for(std::chrono::milliseconds{10});
    if (!process.running())
    {
        throw std::runtime_error("started WorkerProcess should report running");
    }

    const auto deadline = std::chrono::steady_clock::now() + std::chrono::seconds{2};
    while (!handler.handled_validate_request.load() && std::chrono::steady_clock::now() < deadline)
    {
        std::this_thread::sleep_for(std::chrono::milliseconds{10});
    }
    if (!handler.handled_validate_request.load())
    {
        throw std::runtime_error("WorkerProcess did not execute a workflow step");
    }

    process.request_stop();
    if (process.join() != 0)
    {
        throw std::runtime_error("stopped WorkerProcess should join cleanly");
    }
    if (process.running())
    {
        throw std::runtime_error("joined WorkerProcess should not report running");
    }
}

File: tests/bindings/e2e/rust/worker_process_fixture.rs
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;
use std::thread;
use std::time::Duration;

use statespec_generated::json::Json;
use statespec_generated::memory_backend::InMemoryBackend;
use statespec_generated::runtime_workflows::RuntimeWorkflowStore;
use statespec_generated::worker_process::WorkerProcess;
use statespec_generated::worker_process::WorkerProcessConfig;
use statespec_generated::worker_runtime::WorkerRuntime;
use statespec_generated::workflow::{StartWorkflowRequest, WorkflowStore};
use statespec_generated::workflow_provision_service_handlers::ProvisionServiceV1StepHandler;
use statespec_generated::workflow_provision_service_registry;
use statespec_generated::workflow_step_handlers::{
    WorkflowStepHandlerContext, WorkflowStepInvokerMap, WorkflowStepResult,
};

#[derive(Clone)]
struct ProcessWorkflowStepHandler {
    handled_validate_request: Arc<AtomicBool>,
}

impl ProvisionServiceV1StepHandler<InMemoryBackend> for ProcessWorkflowStepHandler {
    fn handle_validate_request(
        &self,
        _tx: &mut <InMemoryBackend as statespec_generated::backend::Backend>::Tx,
        context: &WorkflowStepHandlerContext,
    ) -> statespec_generated::backend::BackendResult<WorkflowStepResult> {
        assert_eq!(context.workflow_name, "ProvisionService");
        assert_eq!(context.step_name, "validate_request");
        self.handled_validate_request.store(true, Ordering::SeqCst);
        Ok(WorkflowStepResult::complete(Some(
            "create_remote_service".to_string(),
        )))
    }

    fn handle_create_remote_service(
        &self,
        _tx: &mut <InMemoryBackend as statespec_generated::backend::Backend>::Tx,
        _context: &WorkflowStepHandlerContext,
    ) -> statespec_generated::backend::BackendResult<WorkflowStepResult> {
        Ok(WorkflowStepResult::complete(Some(
            "wait_for_remote_service".to_string(),
        )))
    }

    fn handle_wait_for_remote_service(
        &self,
        _tx: &mut <InMemoryBackend as statespec_generated::backend::Backend>::Tx,
        _context: &WorkflowStepHandlerContext,
    ) -> statespec_generated::backend::BackendResult<WorkflowStepResult> {
        Ok(WorkflowStepResult::complete(None))
    }
}

#[test]
fn generated_worker_process_lifecycle() {
    let backend = InMemoryBackend::new();
    let runtime = WorkerRuntime::new(backend.clone());
    let handled_validate_request = Arc::new(AtomicBool::new(false));
    let handler = ProcessWorkflowStepHandler {
        handled_validate_request: Arc::clone(&handled_validate_request),
    };
    let mut workflow_step_invokers = WorkflowStepInvokerMap::new();
    workflow_provision_service_registry::register_workflow_step_invokers::<InMemoryBackend>(
        &mut workflow_step_invokers,
        Arc::new(handler),
    );
    let workflow_step_invokers = Arc::new(workflow_step_invokers);
    let config = WorkerProcessConfig {
        bootstrap_on_start: true,
        worker_poll_interval_ms: 1,
    };
    let process = Arc::new(WorkerProcess::with_config(
        runtime,
        workflow_step_invokers,
        config,
    ));

    assert!(
        process.join().is_err(),
        "joining a WorkerProcess before start should fail"
    );
    process.request_stop();

    process
        .runtime
        .bootstrap()
        .expect("bootstrapping Worker runtime failed");
    let workflows = RuntimeWorkflowStore::new();
    workflows
        .start(
            &backend,
            &StartWorkflowRequest {
                workflow_execution_id: "wf-process-1".to_string(),
                workflow_name: "ProvisionService".to_string(),
                workflow_version: 1,
                start_step: "validate_request".to_string(),
                state: Json::Object(Default::default()),
            },
        )
        .expect("starting workflow failed");

    process.start().expect("starting WorkerProcess failed");
    assert!(
        process.start().is_err(),
        "starting WorkerProcess twice should fail"
    );
    thread::sleep(Duration::from_millis(10));
    assert!(
        process.is_running(),
        "started WorkerProcess should report running"
    );

    for _ in 0..200 {
        if handled_validate_request.load(Ordering::SeqCst) {
            break;
        }
        thread::sleep(Duration::from_millis(10));
    }
    assert!(
        handled_validate_request.load(Ordering::SeqCst),
        "WorkerProcess did not execute a workflow step"
    );

    process.request_stop();
    process
        .join()
        .expect("stopped WorkerProcess should join cleanly");
    assert!(
        !process.is_running(),
        "joined WorkerProcess should not report running"
    );
}
